test/test.cpp

```diff
@@ -561,8 +564,90 @@ TEST_CASE("DataBox Interpolation with piecewise grids",
             error);
         REQUIRE(error <= EPSTEST);
       }
+
       // cleanup
       free(db);
+      free(dbh);
+    }
+  }
+}
+
+SCENARIO("Serializing and deserializing a DataBox",
+         "[DataBox][PiecewiseGrid1D][Serialize]") {
+  GIVEN("A piecewise grid") {
+    constexpr int NGRIDS = 2;
+    constexpr Real xmin = 0;
+    constexpr Real xmax = 1;
+
+    RegularGrid1D g1(xmin, 0.35 * (xmax - xmin), 3);
+    RegularGrid1D g2(0.35 * (xmax - xmin), xmax, 4);
+    PiecewiseGrid1D<NGRIDS> g = {{g1, g2}};
+
+    const int NCOARSE = g.nPoints();
+
+    THEN("The piecewise grid contains a number of points equal the sum of "
+         "the points of the individual grids") {
+      REQUIRE(g.nPoints() == g1.nPoints() + g2.nPoints());
+    }
+
+    WHEN("We construct and fill a 3D DataBox based on this grid") {
+      constexpr int RANK = 3;
+      PiecewiseDB<NGRIDS> dbh(Spiner::AllocationTarget::Host, NCOARSE, NCOARSE,
+                              NCOARSE);
+      for (int i = 0; i < RANK; ++i) {
+        dbh.setRange(i, g);
+      }
+      for (int iz = 0; iz < NCOARSE; ++iz) {
+        for (int iy = 0; iy < NCOARSE; ++iy) {
+          for (int ix = 0; ix < NCOARSE; ++ix) {
+            Real x = g.x(ix);
+            Real y = g.x(iy);
+            Real z = g.x(iz);
+            dbh(iz, iy, ix) = linearFunction(z, y, x);
+          }
+        }
+      }
+      WHEN("We serialize the DataBox") {
+        std::size_t serial_size = dbh.serializedSizeInBytes();
+        REQUIRE(serial_size == (sizeof(dbh) + dbh.sizeBytes()));
+
+        char *db_serial = (char *)malloc(serial_size * sizeof(char));
+        std::size_t write_offst = dbh.serialize(db_serial);
+        REQUIRE(write_offst == serial_size);
+
+        THEN("We can initialize a new databox based on the serialized one") {
+          PiecewiseDB<NGRIDS> dbh2;
+          std::size_t read_offst = dbh2.deSerialize(db_serial);
+          REQUIRE(read_offst == write_offst);
+
+          AND_THEN("They do not point to the same memory") {
+            // checks DataBox pointer
+            REQUIRE(dbh2.data() != dbh.data());
+            // checks accessor agrees
+            REQUIRE(&dbh2(0) != &dbh(0));
+          }
+
+          AND_THEN("The shape is correct") {
+            REQUIRE(dbh2.rank() == dbh.rank());
+            REQUIRE(dbh2.size() == dbh.size());
+            for (int d = 1; d <= 3; ++d) {
+              REQUIRE(dbh2.dim(d) == dbh.dim(d));
+            }
+          }
+
+          AND_THEN("The contents are correct") {
+            for (int i = 0; i < dbh.size(); ++i) {
+              REQUIRE(dbh(i) == dbh2(i));
+            }
+          }
```

Comment:
Can you maybe add another test to make sure that two `deSerialize`d  databoxes point to the same memory? Something like
```suggestion
          }
          AND_THEN("Other databoxes initialized in the same way point to shared memory") {
            PiecewiseDB<NGRIDS> dbh3;
            auto new_read_offset = dbh3.deSerialize(db_serial);
            // Both deserialized databoxes point to the same data
            CHECK(dbh3.data() == dbh2.data());
            // Yet they are separate databoxes
            CHECK(&dbh3(0) != &dbh2(0));
          }
```

Reply:
FYI - I am using `CHECK` rather than `REQUIRE` so that if indeed the two test assertions could point to different failure modes, both failure modes would be exposed when the test fails. `REQUIRE`  ends execution at the first failure. I think it's highly unlikely that these would fail separately though, so `REQUIRE` is probably fine.

Reply:
this is a good idea. done. BTW the accessor check should also check they AGREE as that's pointing to the underlying data. But you can check the databoxes themselves are different.

Reply:
Ah okay, I misunderstood. Test looks good!